lib/shelley/exe/shelley-test-cluster.hs

```diff
@@ -0,0 +1,224 @@
+import Cardano.Wallet.Primitive.SyncProgress
+    ( SyncTolerance (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..) )
+import Cardano.Wallet.Shelley
+    ( SomeNetworkDiscriminant (..)
+    , serveWallet
+    , setupTracers
+    , tracerSeverities
+    )
+import Cardano.Wallet.Shelley.Compatibility
+    ( Shelley )
+import Cardano.Wallet.Shelley.Launch
+    ( ClusterLog (..)
+    , RunningNode (..)
+    , defaultPoolConfigs
+    , moveInstantaneousRewardsTo
+    , nodeMinSeverityFromEnv
+    , oneMillionAda
+    , sendFaucetFundsTo
+    , walletListenFromEnv
+    , walletMinSeverityFromEnv
+    , withCluster
+    , withSystemTempDir
+    , withTempDir
+    )
+import Control.Arrow
+    ( first )
+import Control.Monad
+    ( void )
+import Control.Tracer
+    ( traceWith )
+import Data.Proxy
+    ( Proxy (..) )
+import System.IO
+    ( BufferMode (..), hSetBuffering, stdout )
+import Test.Integration.Faucet
+    ( genRewardAccounts, mirMnemonics, shelleyIntegrationTestFunds )
+
+import qualified Data.Text as T
+
+-- |
+-- # OVERVIEW
+--
+-- This starts a cluster of Cardano nodes with:
+--
+-- - 1 relay node
+-- - 1 BFT leader
+-- - 4 stake pools
+--
+-- The BFT leader and pools are all fully connected. The network starts in the
+-- Byron Era and transitions into the Shelley era. Once in the Shelley era and
+-- once pools are registered and up-and-running, an instance of cardano-wallet
+-- is started.
+--
+-- Pools have slightly different settings summarized in the table below:
+--
+-- | #       | Pledge | Retirement      | Metadata       |
+-- | ---     | ---    | ---             | ---            |
+-- | Pool #0 | 2M Ada | Never           | Genesis Pool A |
+-- | Pool #1 | 1M Ada | Epoch 3         | Genesis Pool B |
+-- | Pool #2 | 1M Ada | Epoch 100_000   | Genesis Pool C |
+-- | Pool #3 | 1M Ada | Epoch 1_000_000 | Genesis Pool D |
+--
+-- Pools' metadata are hosted on static local servers started alongside pools.
+--
+-- # PRE-REGISTERED DATA
+--
+-- The cluster also comes with a large number of pre-existing faucet wallets and
+-- special wallets identified by recovery phrases. Pre-registered wallets can be
+-- seen in
+--
+--   `lib/core-integration/src/Test/Integration/Faucet.hs`.
+--
+-- All wallets (Byron, Icarus, Shelley) all have 10 UTxOs worth 100_000 Ada
+-- each (so 1M Ada in total). Additionally, the file also contains a set of
+-- wallets with pre-existing rewards (1M Ada) injected via MIR certificates.
+-- These wallets have the same UTxOs as other faucet wallets.
+--
+-- Some additional wallets of interest:
+--
+-- - (Shelley) Has a pre-registered stake key but no delegation.
+--
+--     [ "over", "decorate", "flock", "badge", "beauty"
+--     , "stamp", "chest", "owner", "excess", "omit"
+--     , "bid", "raccoon", "spin", "reduce", "rival"
+--     ]
+--
+-- - (Shelley) Contains only small coins (but greater than the minUTxOValue)
+--
+--     [ "either" , "flip" , "maple" , "shift" , "dismiss"
```

Comment:
Nice docs, but feels like it might get out of date very easily? Also is a doc comment actually the right place for this information?

Maybe we could add a new `shelley-test-cluster mnemonics`, and change `shelleyIntegrationTestFunds` to also include a `String` description for each wallet?

The output of such a command would be huge. We could add some logic to protect against it, or maybe just something like grep:

E.g.
```bash
$ shelley-test-cluster query-mnemonic "small coins"
shelley/special/seq/Contains only small coins (but greater than the minUTxOValue) 1
    either flip maple shift dismiss bridge sweet reveal green tornado need patient wall stamp pass

shelley/special/seq/Contains only small coins (but greater than the minUTxOValue) 2
    reveal sweet green tornado need patient wall stamp pass either flip maple shift dismiss bridge

```

Reply:
> Nice docs, but feels like it might get out of date very easily? Also is a doc comment actually the right place for this information?

I wondered the same and actually considered the same idea you're proposing here. Although arguably: 

- we haven't changed those mnemonics in several months and have no reason to change them in the upcoming future. 
- that is for now, mostly an internal tool, so if this end up outdated, we'll most likely notice it quite quickly. 

So, the size of the problem seems not large enough to spend effort on making this better :sweat_smile: ... arguably.

Reply:
I mean, the same argument applies for the pools configurations and even the cluster setup (what if we add a new node tomorrow? the doc becomes outdated). All this doc describes implementation details, so by its very nature, it's fragile.

Reply:
I remember having to re-regenerate and update faucet mnemonics several for Shelley, which was not fun at all. With the HFC integration we finally got it much more automated.

I think just adding a command to output all mnemonics with a description, requiring manual `| grep ""` piping, would have been really simple.

But sure, it's perhaps unlikely they'll change now. And perhaps still best to keep it comment, for now.

Btw, I wonder if this could be used for ADP-381 eventually.

Reply:
Perhaps ultimately, we can clean-up our integration tests setup to use that executable. Spawn it in a process and go ahead with the rest of the tests. Right now the integration setup is very complicated with contexts and mvars a bit everywhere waiting for stuff.